### PR TITLE
Flexible sidebar

### PR DESCRIPTION
**Pull request by jamesscottbrown**
**What does this change?**

**Why?**

**How?**

**Related issues**:

**Does this introduce new dependencies?**

**How is it tested?**

**How is it documented?**

**Are light and dark themes considered?**

**Is it complete?**

- [ ] Have you included changeset file?
- [ ] If this adds a new component, is it exported via `index.js`?

### CODE REVIEW

**Comment by jamesscottbrown** [comment]
This enables sidebars containing tab icons.

It has two types of sidebar headings: `SidebarAppHeader` for app headings with the blue flash, and `SidebarHeader` that matches the header in Figma.

image


image

**Comment by jamesscottbrown** [comment]
Preview: https://dev.ldn-gis.co.uk/storybook-flexible-sidebar/?path=/docs/ldn-viz-introduction--documentation

Preview of the components that go inside the sidebar: https://dev.ldn-gis.co.uk/storybook-flexible-sidebar/?path=/story/ui-sidebarelements-sidebarappheader--with-description-and-more-info

File: packages/ui/src/lib/sidebar/Sidebar.svelte [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+	const appShell = getContext('appShell');
+	$: isAlignedRight = $appShell.isAlignedRight();
+</script>
+
+<section
+	class="relative bg-core-grey-800 w-full h-full flex flex-col text-white text-sm z-30 {classes}"
+>
+	<div
+		class="absolute"
```
Comment (jamesscottbrown) [inline packages/ui/src/lib/sidebar/Sidebar.svelte:18]:
```suggestion
		class="absolute dark"
```

I think we need to apply the `dark` class here, or else components like `<Select>` used in the sidebar use the wrong colors

File: packages/ui/src/lib/sidebar/SidebarLeft.svelte [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	$: isAlignedLeft = $appShell.isAlignedLeft();
+
+	$: console.log('width:', $appShell.sidebarWidth);
+</script>
+
+<section
+	class="relative bg-core-grey-800 w-full h-full flex flex-col text-white text-sm z-30 {classes}"
+>
+	<div
+		class="absolute"
```
Comment (jamesscottbrown) [inline packages/ui/src/lib/sidebar/SidebarLeft.svelte:20]:
```suggestion
		class="absolute dark"
```

I think we need to apply the `dark` class here, or else components like `<Select>` used in the sidebar use the wrong colors

File: packages/ui/src/lib/appShell/AppShell.svelte [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	$: $isAlwaysOpen = aoProp;
+	$: $sidebarPlacementStore = bpProp;
+
+	setContext('sidebarAlwaysOpen', isAlwaysOpen);
+	setContext('sidebarIsOpen', isOpen);
+	setContext('sidebarPush', sidebarPush);
+	setContext('sidebarPlacement', sidebarPlacementStore);
+</script>
+
+<!-- Inorder to get consistent width between code and css we need to use the innerwidth of the window -->
```
Comment (jamesscottbrown) [inline packages/ui/src/lib/appShell/AppShell.svelte:61]:
```suggestion
<!-- In order to get consistent width between code and css we need to use the innerWidth of the window -->
```

File: packages/ui/src/lib/appShell/AppShell.svelte [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+					Provide some main content. The main cotent you provide should have appropriate padding
+					applied...
+				</span>
+			</p>
+		</slot>
+	</main>
+
+	<slot name="sidebar">Sidebar</slot>
+
+	<!-- This div exists to push content to the side of the sidebar	when sidebar push is set to true-->
```
Comment (jamesscottbrown) [inline packages/ui/src/lib/appShell/AppShell.svelte:78]:
```suggestion
	<!-- This div exists to push content to the side of the sidebar	when sidebarPush is set to true-->
```

File: packages/ui/src/lib/sidebar/Sidebar.svelte [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	$: tabPlacementClasses = tabPlacementLookup[placement];
+	$: widthClasses = widthLookup[width][placement];
+	$: heightClasses = heightLookup[width][placement];
+
+	// set a store containing the width of the sidebar (for use in app shell and elsewhere up the tree)
+	$: $sidebarWidthStore = width;
+</script>
+
+<div class={classNames(wrapperClasses, placementClasses)}>
+	<!-- $sidebarIsOpen ? widthClasses : '' -->
```
Comment (jamesscottbrown) [inline packages/ui/src/lib/sidebar/Sidebar.svelte:43]:
```suggestion
```

**Comment by mikeldn** [comment]
Horizontal Siderbar Tabs in Storybook docs, aren't showing as Horizontal.
Also, Icon placement within tabs is different between Horz & Vert Tabs. 

image

**Comment by mikeldn** [comment]
Sidebar toggle (custom icon) not showing a different icon in docs, so a bit confusing?!?

image

**Comment by ChrisKnightLDN** [comment]
> * Horizontal Siderbar Tabs in Storybook docs, aren't showing as Horizontal.
> * Selected tab state not correct by default in strory
> * Icon placement within tabs is different between Horz & Vert Tabs.

This was due to the use of context within the component. I have updated the story to use a context wrapper to reflect this.

**Comment by ChrisKnightLDN** [comment]
> Sidebar toggle (custom icon) not showing a different icon in docs, so a bit confusing?!?

This was also due to context that is provided by the sidebar component - therefore absent in the story where the component is isolated. I have updated the story to provide the context that the functionality requires.

**Pull request merged**